### ISSUE TITLE
using-pandoc-to-read-docx-to-capture-contents-of-a-docx-textbox

### ISSUE THREAD

**Issue by intotecho** [jgm/pandoc#3086]
Referring to question
[http://stackoverflow.com/questions/39091419/using-pandoc-to-read-docx-to-capture-contents-of-a-docx-textbox]

I am investigating using Pandoc to convert thousands of word docx documents to docbook format. 

I am using a Microsoft converter to convert word97 docs to docx.

Here is a sample document. The text boxes on the left hand margin are not retained when reading the docx. 
[32_07s01_bany.docx](https://github.com/jgm/pandoc/files/433808/32_07s01_bany.docx)

These documents all use a text box to record the provenance and change history of each paragraph.

I would like to store the textbox content as a metadata attribute in the docbook. Then it can be rendered as desired when the docbook is converted to PDF or HTML.

Thanks for having a look. I agree it may be difficult to associate the textbox to the paragraph which is not in the text box.

**Comment by jkr** [comment]
It would be easy enough to get text out of the text box, but the question would be where to put it. It doesn't make sense to me to put it into the metadata -- that's not something that we normally do with text (unless it actually is metadata, like a title or author-name). I see three options:

1. Just stick it into the document, as a paragraph, where it appears. That is, where it appears in the xml -- which might be different from where it appears geometrically on the page.

2. Stick it in the caption of an empty image. This has some feel of consistency, since a textbox is part of  a `<w:pict>` element. But it also is weird, because I don't think we do that anywhere else.

3. Continue to ignore it.

There's actually a fourth option here, which would be to include it in a special div which you could filter on. But we could *do* that for lots of info in docx---underlines vs. emph, endnotes vs. footnotes---and we don't, since it ends up being a weird ad-hoc extension of the pandoc AST. (We do it with track-changes, but that seemed important enough to the way people use docx.)

So the answer is, it wouldn't be hard to do something with it, but I have to give some thought to what, exactly. @jgm -- do you have any thoughts on this?

**Comment by intotecho** [comment]
The special div sounds like a good approach for my application because the renderer could filter it out or display it according to what makes sense to the publisher. Thanks for looking into it.

**Comment by awgeorge** [comment]
We use textboxes in .docx in order to create code snippets. This way we can visually separate the sample and set a custom background color to better reflect the syntax highlighting. It's therefore important for us to at least read the textbox - otherwise all of our sample are lost. The ultimate scenario would be to convert them to code blocks (```) - But that wouldn't help @intotecho.

I guess my main point is, textbooks usually contain important information - and losing the content is not ideal.

**Comment by jgm** [comment]
@jkr I think your option 1 would be fine, and putting it in a special div is a good idea, since that would allow people to float it if needed (perhaps with a filter).

**Comment by Zeppelin456** [comment]
Most doc converters add text box content and header and footer content at either the end of the document or at the start of the document in one big block.   Then people know that this is stuff they need to move around.   At least the content then gets extracted and it is easy to see which parts need to be moved around.  

I use this to convert text from job applicant cvs who have sent to us, just into plain text and not even to markdown, however the most important information (their name, contact details etc, are usually in the header of the doc, and sometimes text boxes.   If I can at least be sure the information is there, even if in the wrong place then I would be overjoyed!!   End of document or top is fine for me.  Maybe a switch --worddoc-headers and set to doc-end, doc-start or in-place.  and a similar one for text boxes?   Same for .docs as well as .docx, but less important for me as less used these days.

**Comment by dtlhlbs** [comment]
This would be super useful. I have a Word doc containing code and other example content in text boxes and it's all lost on conversion.

**Comment by githubyouser** [comment]
Yes **_please_**, we need to be able to "keep" the text in boxes somehow! I have to manually go through and add that text back in after a conversion. :(

**Comment by dtlhlbs** [comment]
As a workaround I run this C# code to extract textbox content and place it where the textbox anchor is. It's pretty rough and sometimes the Word document is left open for some reason and i have to reboot. After `pandoc` runs I then parse and handle content surrounded by the `TEXTBOX START%%%` and `%%%TEXTBOX END` markup.

```cs
﻿using Microsoft.Office.Interop.Word;
using System.Text.RegularExpressions;

string path = args[0];

Console.WriteLine($"Parsing {path}");

Application word = new Application();
Document doc = word.Documents.Open(path);

try
{

    foreach (Shape shp in doc.Shapes)
    {
        if (shp.TextFrame.HasText != 0)
        {
            string text = shp.TextFrame.TextRange.Text;
            int page = shp.Anchor.Information[WdInformation.wdActiveEndPageNumber];
            string summary = Regex.Replace(text, @"\r\n?|\n", " ");
            Console.WriteLine($"++++textbox++++ Page {page}: {summary.Substring(0, Math.Min(summary.Length, 40))}");

            string newtext = @$"{Environment.NewLine}TEXTBOX START%%%{text}%%%TEXTBOX END{Environment.NewLine}";

            var range = shp.Anchor;
            range.InsertBefore(Environment.NewLine);
            range.Collapse();
            range.Text = newtext;
            range.set_Style(WdBuiltinStyle.wdStyleNormal);
        }
    }

    string newFile = Path.GetFullPath(path) + ".transformed.docx";
    doc.SaveAs2(newFile);
}
finally
{
    doc.Close();
    word.Quit();
}
```

**Comment by githubyouser** [comment]
Thanks, @dtlhlbs! Sadly, I don't know the first thing about C#, so that doesn't help me much. I have worked with VBA... maybe there's a similar script in VBA?

**Comment by dtlhlbs** [comment]
@githubyouser I actually started with some VBA code I found on the web. I don't recall the source, but this was the VBA I was testing before I switched to C#. It wouldn't be hard to find the VBA equivalent of my final C#.

```vb
Sub ConvertTextboxes()
    Dim myTextBox As Shape
    Dim page As Integer
    For Each myTextBox In ActiveDocument.Shapes
      If myTextBox.Type = msoTextBox Then
        myTextBox.Select
        page = Selection.Information(wdActiveEndPageNumber)
        Debug.Print "****************************"
        Debug.Print "Page " & page
        Debug.Print myTextBox.TextFrame.TextRange.Text
      End If
    Next
End Sub

// get the anchor paragraph, we could use this to position text
Dim oRngAnchor As Range
Set oRngAnchor = shp.Anchor.Paragraphs(1).Range
oRngAnchor.InsertBefore "*TEXTBOX HERE*"
```

**Comment by githubyouser** [comment]
Thanks, @dtlhlbs, that pointed me in the right direction. I managed to create a [VBA script](https://gist.github.com/githubyouser/28e8f56ac59697abeee4a05d88afd94d) that seems to do the trick. 

It would still be a lot nicer to have Pandoc do it by default though. :(
